### PR TITLE
feat: add vk and yandex providers to oidc providers and documentation

### PR DESCRIPTION
**Pull request by aenoralanar**
## Related issue

#1234

## Proposed changes

- Adds provider_vk.go and updates the configuration schema to allow for vk as an OIDC provider
- Adds provider_yandex.go and updates the configuration schema to allow for yandex as an OIDC provider
- Adds description to documentation

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
Could you please enable [write access for maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)? 🤔 

Thank you! 🙏

File: go.mod [hunk 1]
```diff
@@ -97,4 +97,5 @@ require (
 	golang.org/x/oauth2 v0.0.0-20201208152858-08078c50e5b5
 	golang.org/x/sync v0.0.0-20201020160332-67f06af15bc9
 	golang.org/x/tools v0.1.0
+	github.com/davecgh/go-spew v1.1.1
```
Comment (aeneasr) [inline go.mod:100]: Looks like this was added for debugging - can we remove it?

File: selfservice/strategy/oidc/provider_vk.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+func (g *ProviderVK) Claims(ctx context.Context, exchange *oauth2.Token) (*Claims, error) {
+
+	o, err := g.OAuth2(ctx)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	client := o.Client(ctx, exchange)
+
+	u, err := url.Parse("https://api.vk.com/method/users.get?fields=photo_200,nickname,bdate,sex&access_token=" + exchange.AccessToken + "&v=5.103")
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_vk.go:65]: Looks like these fields are something people would like to configure themselves?

File: selfservice/strategy/oidc/provider_vk.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	defer resp.Body.Close()
+
+	type User struct {
+		Id            int    `json:"id,omitempty"`
+		FirstName     string `json:"first_name,omitempty"`
+		LastName      string `json:"last_name,omitempty"`
+		Nickname      string `json:"nickname,omitempty"`
+		Picture       string `json:"photo_200,omitempty"`
+		Email         string
+		EmailVerified bool
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_vk.go:88]: These two fields are missing JSON annotations - also is this the complete list of claims and where could we find the complete list of fields in the VK docs?

File: selfservice/strategy/oidc/provider_yandex.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+	defer resp.Body.Close()
+
+	var user struct {
+		Id            string `json:"id,omitempty"`
+		FirstName     string `json:"first_name,omitempty"`
+		LastName      string `json:"last_name,omitempty"`
+		Email         string `json:"default_email,omitempty"`
+		EmailVerified bool
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_yandex.go:83]: This field is missing JSON annotations - also is this the complete list of claims and where could we find the complete list of fields in the Yandex docs?

File: selfservice/strategy/oidc/provider_yandex.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+		Gender        string `json:"sex,omitempty"`
+		BirthDay      string `json:"birthday,omitempty"`
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	if !user.PictureEmpty {
+		user.Picture = "https://avatars.yandex.net/get-yapic/" + user.Picture + "/islands-200"
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_yandex.go:95]: What is `islands-200` for?

**Review by aeneasr — COMMENTED** [review]
Great job! This looks almost good to merge! To be able to merge it could you please address the comments in the review? Also, it would be great to see a short screen cast of the login flow for VK and Yandex to confirm that everything is working as it should - given that we currently do not have e2e testing for external providers? Thank you 🙏  😄

File: selfservice/strategy/oidc/provider_yandex.go (hunk 5 above)
Comment (NickUfer) [reply selfservice/strategy/oidc/provider_yandex.go:95]: Seems to be the avatar size, 200x200 pixels.

https://i.imgur.com/aqEZagw.png

File: selfservice/strategy/oidc/provider_vk.go (hunk 3 above)
Comment (NickUfer) [reply selfservice/strategy/oidc/provider_vk.go:88]: api method: https://vk.com/dev/users.get
scopes: https://vk.com/dev/permissions
claims: https://vk.com/dev/fields
more claims: https://vk.com/dev/fields_2

File: selfservice/strategy/oidc/provider_yandex.go (hunk 4 above)
Comment (aenoralanar) [reply selfservice/strategy/oidc/provider_yandex.go:83]: api: https://yandex.com/dev/passport/doc/dg/reference/request.html
claims: https://yandex.com/dev/passport/doc/dg/reference/response.html

**Comment by aenoralanar** [comment]
Sorry for my late response.
Added write access for @aeneasr and @NickUfer.
Screencast:
scrkr

File: selfservice/strategy/oidc/provider_vk.go (hunk 2 above)
Comment (aenoralanar) [reply selfservice/strategy/oidc/provider_vk.go:65]: Most of the fields do not match claims.
https://vk.com/dev/fields 
https://www.ory.sh/kratos/docs/concepts/credentials/openid-connect-oidc-oauth2/#external-variable-claims

File: go.mod (hunk 1 above)
Comment (aenoralanar) [reply go.mod]: Already removed.

File: selfservice/strategy/oidc/provider_yandex.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	} else {
+		user.Picture = ""
+	}
+
+	if user.Email != "" {
+		user.EmailVerified = true
+	}
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_yandex.go:102]: I couldn't find this in the docs. Is a set email address really verified? If so, could you please add a code-doc that references the section in the Yandex docs here? Thank you! :)

File: selfservice/strategy/oidc/provider_vk.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	user := response.Result[0]
+
+	if email := exchange.Extra("email"); email != nil {
+		user.Email = email.(string)
+		user.EmailVerified = true
+	}
```
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_vk.go:106]: Same question as with Yandex:

> I couldn't find this in the docs. Is a set email address really verified? If so, could you please add a code-doc that references the section in the VK docs here? Thank you! :)

Comment (aenoralanar) [reply selfservice/strategy/oidc/provider_vk.go:106]: i don't find ref in VK documentation. Email linking process requires confirmation(tested).

File: selfservice/strategy/oidc/provider_yandex.go (hunk 6 above)
Comment (aenoralanar) [reply selfservice/strategy/oidc/provider_yandex.go:102]: I don't find ref in Yandex documentation. Yandex uses its own email(@yandex or custom domain) as default_email.

**Comment by TorinAsakura** [comment]
@aeneasr hey
and so?

File: selfservice/strategy/oidc/provider_yandex.go (hunk 6 above)
Comment (aeneasr) [reply selfservice/strategy/oidc/provider_yandex.go:102]: Ok, so does that mean that the email set in yandex is not verified? Or is it verified? Is the email always linked to the yandex account? Setting a verified email in OIDC is pretty dangerous, so we should only do this if we are certain that the email is indeed verified!

File: selfservice/strategy/oidc/provider_vk.go (hunk 7 above)
Comment (aeneasr) [reply selfservice/strategy/oidc/provider_vk.go:106]: Same thing here, automatically setting email verified to true has some implications here. If we're not sure if the email was actually verified, this should be false!

**Comment by aeneasr** [comment]
Hey there, currently on vacation so responses and reviews take a bit longer ;)

**Review by aeneasr — APPROVED** [review]

**Comment by aeneasr** [comment]
If you rebase / merge master test should pass!

**Comment by aeneasr** [comment]
I tried pushing some changes required for merging the PR to your fork & branch, but it appears that I am not allowed to do so 😕 

```
% git push ...
ERROR: Permission to push denied to aeneasr.
fatal: could not read from the remote repository.

Please make sure that you have the correct access rights
and the repository exists.
```

But the good news is, giving access is easy! :relaxed: All you need to do is [enable write access for maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork). Thank you! 😄

**Comment by aenoralanar** [comment]
> But the good news is, giving access is easy! relaxed All you need to do is enable write access for maintainers. Thank you! smile

Doesn't work on organisation repositories(isaacs/github#1681). 
I sent an invite - https://github.com/atls/kratos/invitations.

**Comment by TorinAsakura** [comment]
> I tried pushing some changes required for merging the PR to your fork & branch, but it appears that I am not allowed to do so 😕
> 
> ```
> % git push ...
> ERROR: Permission to push denied to aeneasr.
> fatal: could not read from the remote repository.
> 
> Please make sure that you have the correct access rights
> and the repository exists.
> ```
> 
> But the good news is, giving access is easy! ☺️ All you need to do is [enable write access for maintainers](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork). Thank you! 😄

Hey!
And so?

**Review by aeneasr — APPROVED** [review]
Thank you very much for your hard work and sorry for the long turnaround time!

**Pull request merged**